framework/integration/test/reconciliation/reconciliation_test.go

```diff
@@ -1,21 +1,113 @@
-// +build k8srequired
-
 package reconciliation
 
 import (
+	"reflect"
 	"testing"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 
 	"github.com/giantswarm/operatorkit/framework/integration/client"
 )
 
 // Test_Finalizer_Integration_Reconciliation is a integration test for
 // the proper replay and reconciliation of delete events with finalizers.
 func Test_Finalizer_Integration_Reconciliation(t *testing.T) {
+	configMapName := "test-cm"
+	testFinalizer := "operatorkit.giantswarm.io/test-operator"
 	testNamespace := "finalizer-integration-reconciliation-test"
+	testOtherFinalizer := "operatorkit.giantswarm.io/other-operator"
+	operatorName := "test-operator"
 
 	client.MustSetup(testNamespace)
 	defer client.MustTeardown(testNamespace)
 
-	// TODO: Implement the actual test here.
+	operatorkitFramework, err := client.NewFramework(operatorName, testNamespace)
+	if err != nil {
+		t.Fatal("expected", nil, "got", err)
+	}
+
+	// We create a configmap, but add a finalizer of another operator. This will
+	// cause the configmap to continue existing after the framework removes it own
+	// finalizer.
+	cm := &corev1.ConfigMap{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: "v1",
+			Kind:       "ConfigMap",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      configMapName,
+			Namespace: testNamespace,
+			Finalizers: []string{
+				testOtherFinalizer,
+			},
+		},
+		Data: map[string]string{},
+	}
+	_, err = client.CreateConfigMap(cm)
+	if err != nil {
+		t.Fatal("expected", nil, "got", err)
+	}
+
+	// We start the framework.
+	go operatorkitFramework.Boot()
+
+	// We wait the absolute maximum amount of time here:
+	// 20 second ResyncPeriod + 2 second RateWait + 3 second for safety.
+	time.Sleep(25 * time.Second)
+
+	// We get the ConfigMap after the framework has been started.
+	resultConfigMap, err := client.GetConfigMap(configMapName, testNamespace)
+	if err != nil {
+		t.Fatal("expected", nil, "got", err)
+	}
+
+	// We verify, that the DeletionTimestamp has not been set.
+	if resultConfigMap.GetDeletionTimestamp() != nil {
+		t.Fatalf("DeletionTimestamp != nil, want nil")
+	}
+
+	// We define which finalizers we currently expect.
+	expectedFinalizers := []string{
+		testOtherFinalizer,
+		testFinalizer,
+	}
+
+	// We verify, that our finalizer is still set.
+	if !reflect.DeepEqual(resultConfigMap.GetFinalizers(), expectedFinalizers) {
+		t.Fatalf("finalizers == %v, want %v", resultConfigMap.GetFinalizers(), expectedFinalizers)
+	}
+
+	// We delete the configmap now.
+	err = client.DeleteConfigMap(configMapName, testNamespace)
+	if err != nil {
+		t.Fatal("expected", nil, "got", err)
+	}
+
+	// We wait the absolute maximum amount of time here:
+	// 20 second ResyncPeriod + 2 second RateWait + 3 second for safety.
+	time.Sleep(25 * time.Second)
```

Comment:
This is the point, where replay on the delete event happens. I can see this in the log output. 
The object still exists because we added another finalizer at the very beginning. The deletion will therefor replay until that other finalizer is removed, because the object cannot be deleted until then.

I can't think of a way currently to verify that the delete event is replayed.

Reply:
I don't quite follow here, how could I retrieve the value from the counter?

I guess the resource could add it as a label to the object or something?

Reply:
You implement a new resource. This is fully under your control. You configure a framework with it and execute your test. After the test you check your resource and what it did.

Reply:
Yes, I understand that. I am confused because I run the framework like this:
https://github.com/giantswarm/operatorkit/blob/a7025edeb543ae64605ae7bb972db54a06e90050/framework/integration/test/reconciliation/reconciliation_test.go#L56
So the value of the counter would be modified in another goroutine and there is no channel so I won't be able to simply access it.Or am I totally misunderstanding here?

Reply:
Your test is synchronous at the end even if you run multiple goroutines. Your resource is ideally a pointer so it can increment a counter by retaining the updated values which you can inspect at the end of your tests.

Reply:
Oh I see, that makes total sense. My bad.

Reply:
I do this now, can you check if this is makes sense to you?